src/librustc_mir/borrow_check.rs

```diff
@@ -642,29 +642,143 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
         let lvalue = self.base_path(lvalue_span.0);
 
         let maybe_uninits = &flow_state.uninits;
-        let move_data = maybe_uninits.base_results.operator().move_data();
-        if let Some(mpi) = self.move_path_for_lvalue(context, move_data, lvalue) {
+
+        // Bad scenarios:
+        //
+        // 1. Move of `a.b.c`, use of `a.b.c`
+        // 2. Move of `a.b.c`, use of `a.b.c.d` (without first reinitializing `a.b.c.d`)
+        // 3. Move of `a.b.c`, use of `a` or `a.b`
+        // 4. Uninitialized `(a.b.c: &_)`, use of `*a.b.c`; note that with
+        //    partial initialization support, one might have `a.x`
+        //    initialized but not `a.b`.
+        //
+        // OK scenarios:
+        //
+        // 5. Move of `a.b.c`, use of `a.b.d`
+        // 6. Uninitialized `a.x`, initialized `a.b`, use of `a.b`
+        // 7. Copied `(a.b: &_)`, use of `*(a.b).c`; note that `a.b`
+        //    must have been initialized for the use to be sound.
+        // 8. Move of `a.b.c` then reinit of `a.b.c.d`, use of `a.b.c.d`
+
+
+        // The dataflow tracks the shallow prefixes distinctly, in
+        // order to track substructure initialization separately from
+        // the whole structure.
+        //
+        // Therefore, it suffices to find the longest shallow prefix
+        // for which we have a MovePath (if any); the dataflow will
+        // take care of updating its state in response to
+        // super-structure moves and reinitializations.
+        //
+        // This is scenarios 1 and 2 above.
+
+        debug!("check_if_path_is_moved part1 lvalue: {:?}", lvalue);
+        match self.move_path_closest_to(lvalue, PrefixSet::Shallow) {
```

Comment:
Why is this considering only the shallow set of prefixes? This later does a `max_deref_free_locally_rooted_prefix_of` to consider deep prefixes, so I don't see why this is split into 2 parts.